Fix the Github Action CI for Python tests

remrama opened the pull request:
This PR corrects a typo in the Python tests Github Action CI yaml file that was preventing the tests from running on PRs. The pingouin master branch was recently renamed from `master` to `main`, and now the yaml file reflects that. This was identified in [a comment](https://github.com/raphaelvallat/pingouin/pull/443#issuecomment-2400153386) in #443.

remrama commented:
Ok, took a few adjustments but all tests are passing now! Sorry for all the pushes, I still am not aware of a way to test GH Actions locally... I am pushing just to see if something works 😕  If anyone has advice on how to test locally and avoid this approach, please let me know!

### Updated list of changes

I had to bump a few action versions up where the older ones caused an issue, and updated the docs requirements installation to take advantage of the modern python packaging layout that happened recently in #406.

- [x] `[master, develop]` --> `[main]`
- [x] `actions/upload-artifact@v2` --> `actions/upload-artifact@v4` (see [migration guide](https://github.blog/news-insights/product-news/get-started-with-v4-of-github-actions-artifacts/
))
- [x] `actions/checkout@v1` --> `actions/checkout@v4`
- [x] `actions/setup-python@v1` --> `actions/setup-python@v5`
- [x] pip install docs requirements from .toml file with `pip install .[docs]`
- [ ] Pingouin currently uses `codecov/codecov-action@v1` but latest is `v5`. The current version is not leading to any errors, but it leads to warnings [about node usage](https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default/). If you want, I can also try bumping the codevoc action while I'm here. I might run into some token access issues but could play with it.
- [ ] Is there supposed to be a `download-artifact` action in the Python tests workflow? The Pingouin documentation page has an ["Inspect on Github" section](https://pingouin-stats.org/build/html/contributing.html#inspect-on-github) with instructions to view the artifact but they don't correspond to what I see on my end (i.e., I don't see downloadable artifacts to inspect). Maybe I'm missing something here? Also that section refers to tests on Python version 3.8 which Pinguoin no longer runs (I suspect this should be 3.9 since that is the version that builds docs during tests).

Any thoughts on those last 2 hanging points?

raphaelvallat commented:
You're on a roll, thank you!!

I think they changed the interface for downloading artefacts. The artefacts can be found at the bottom of the GH Actions Summary page:

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/c748dcdc-1e67-4fe5-b220-faafc185320d">

You can also find the url in the raw log:

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/5d8f41c8-9861-4fb1-9386-f71fd1563756">

Please do try to update codecov, and let me know if you run into any issues.

raphaelvallat commented on .github/workflows/python_tests.yml:
```diff
@@ -20,17 +20,18 @@ jobs:
       run: |
         python -m pip install --upgrade pip
         pip install .[test]
+        pip install .[docs]
```
Yes please!

remrama commented:
Alright that should be good @raphaelvallat! I also removed the platform and python version from the docs artifact. I added it before, but it's only necessary if you're uploading artifacts from multiple platforms/versions. That's not the case here, so it's cleaner as it was without.

I don't see the same web interface for artifacts as you do in the screenshot, but I found the URL in the logs as you suggested (thanks for the pics, very helpful). Agreed, not a big deal for now.

raphaelvallat reviewed (approved):
Amazing thank you! Merging now
